python_modules/dagster/dagster/_core/definitions/asset_graph_differ.py

```diff
@@ -28,6 +30,71 @@ class AssetDefinitionChangeType(Enum):
     REMOVED = "REMOVED"
 
 
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailNew:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.NEW
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailCodeVersion:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.CODE_VERSION
+    base_value: Optional[str]
+    branch_value: Optional[str]
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailDependencies:
+    """Represents the detailed information about the specific changes between base and branch assets."""
+
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.DEPENDENCIES
+    base_value: Optional[AbstractSet["AssetKey"]]
+    branch_value: Optional[AbstractSet["AssetKey"]]
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailPartitionsDefinition:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.PARTITIONS_DEFINITION
+    base_value: Optional[str]
+    branch_value: Optional[str]
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailTags:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.TAGS
+    base_value: Mapping[str, str]
+    branch_value: Mapping[str, str]
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailMetadata:
```

Comment:
Metadata values can be pretty large, so for this case in particular I think it would be good to just track which keys were changed. If you can do the list of keys changed (or even better, added/changed/removed) that's sufficient.